snapcraft/__init__.py

```diff
@@ -183,23 +159,12 @@ def __init__(self, name, options):
 
     # The API
     def pull(self):
-        """Pull the source code and/or internal prereqs to build the part.
-
-        By default, the base implementation for pull will use the following
-        part properties to retrieve source code:
-
-        - source
-        - source-branch
-        - source-tag
-        - source-type
-
-        If source is empty or does not exist, the phase will be skipped.
+        """Pull any code and/or internal prereqs to build the part.
 
         Override or inherit from this method if you need to implement or
         enhance with custom pull logic.
         """
-        if getattr(self.options, 'source', None):
-            sources.get(self.sourcedir, self.builddir, self.options)
+        pass
```

Comment:
That doesn't sound bad, but as I understood the idea of the lifecycle is something that manages the packaging steps. And a plugin could extend or modify the lifecycle. With implicit steps, then you are constrained in some parts.
Here we have a step called pull that doesn't pull anything. And a step of unknown name because it's internal and won't be published in the docs that does the pull. Confusing.